File: 2023/day4/main.js
import fs from "fs";

var inputFilename = "input";
var input = fs.readFileSync(inputFilename).toString().split("\n");

const main = (values) => {
  const points = values.reduce((acc, line) => {
    const [, allNums] = line.split(":");
    let [winningNums, nums] = allNums.split("|").map((s) => s.trim());
    winningNums = winningNums.split(" ").map((n) => Number(n));
    nums = nums
      .split(" ")
      .filter((n) => Boolean(n))
      .map((n) => Number(n));
    const matches = [...new Set(nums.filter((n) => winningNums.includes(n)))];
    let pts = 0;

    if (matches.length === 1) {
      pts = 1;
    } else if (matches.length !== 0) {
      pts = Math.pow(2, matches.length - 1);
    }

    return acc + pts;
  }, 0);

  const totals = values.map(() => 1);
  const totalCards = values.reduce((acc, line, idx) => {
    const [, allNums] = line.split(":");
    let [winningNums, nums] = allNums.split("|").map((s) => s.trim());
    winningNums = winningNums.split(" ").map((n) => Number(n));
    nums = nums
      .split(" ")
      .filter((n) => Boolean(n))
      .map((n) => Number(n));
    const matches = [...new Set(nums.filter((n) => winningNums.includes(n)))];
    const total = totals[idx];

    for (let i = 1; i <= matches.length; i++) {
      totals[idx + i] += total;
    }

    return acc + total;
  }, 0);

  console.log("Points: ", points);
  console.log("Total Cards: ", totalCards);
};

main(input);

File: 2023/day4/main.rs
use std::{
    fs::File,
    io::{BufRead, BufReader},
};

fn main() {
    let file = File::open("input").expect("Can't open file");
    let buf = BufReader::new(file);

    let lines: Vec<String> = buf
        .lines()
        .map(|l| l.expect("Could not parse line"))
        .collect::<Vec<String>>();

    let mut points = 0;
    let mut totals = vec![1; lines.len()];

    for (idx, line) in lines.iter().enumerate() {
        let parts = line.split(": ").collect::<Vec<&str>>();
        let all_nums = parts[1];
        let num_parts = all_nums.split(" | ").collect::<Vec<&str>>();
        let winning_nums = num_parts[0].split(" ").collect::<Vec<&str>>();
        let nums = num_parts[1].split(" ").collect::<Vec<&str>>();

        let mut winners = nums
            .iter()
            .filter(|&n| winning_nums.contains(n) && !n.is_empty())
            .cloned()
            .collect::<Vec<&str>>();

        winners.sort();
        winners.dedup();

        if winners.len() > 0 {
            let p = i32::pow(2, (winners.len() - 1) as u32);
            points += i32::pow(2, (winners.len() - 1) as u32);
        }

        let total = totals[idx];

        for i in 1..winners.len() + 1 {
            totals[idx + i] += total;
        }
    }

    let total_cards = totals.into_iter().map(|t| t as i32).sum::<i32>();

    println!("Points: {:?}", points);
    println!("Totals cards: {:?}", total_cards);
}
